xcm/pallet-xcm/src/lib.rs

```diff
@@ -1095,6 +1096,48 @@ pub mod pallet {
 /// The maximum number of distinct assets allowed to be transferred in a single helper extrinsic.
 const MAX_ASSETS_FOR_TRANSFER: usize = 2;
 
+impl<T: Config> XcmQueryHandler for Pallet<T> {
+	type QueryId = u64;
+	type BlockNumber = T::BlockNumber;
+	type Error = XcmError;
+
+	/// To check the status of the query, use `fn query()` passing the resultant `QueryId`
+	/// value.
+	fn report_outcome(
+		message: &mut Xcm<()>,
+		responder: impl Into<MultiLocation>,
+		timeout: Self::BlockNumber,
+	) -> Result<Self::QueryId, Self::Error> {
+		let responder = responder.into();
+		let destination = T::UniversalLocation::get()
+			.invert_target(&responder)
+			.map_err(|()| XcmError::LocationNotInvertible)?;
+		let query_id = Self::new_query(responder, timeout, Here);
+		let response_info = QueryResponseInfo { destination, query_id, max_weight: Weight::zero() };
+		let report_error = Xcm(vec![ReportError(response_info)]);
+		message.0.insert(0, SetAppendix(report_error));
+		Ok(query_id)
+	}
+
+	fn take_response(query_id: Self::QueryId) -> QueryResponseStatus<Self::BlockNumber> {
+		match Queries::<T>::get(query_id) {
+			Some(QueryStatus::Ready { response, at }) => {
+				let response = response.try_into();
+				match response {
+					Ok(response) => {
+						Queries::<T>::remove(query_id);
+						Self::deposit_event(Event::ResponseTaken(query_id));
+						QueryResponseStatus::Finished { response, at }
+					},
+					Err(_) => QueryResponseStatus::NotFound, // Found unexpected version
+				}
+			},
+			Some(QueryStatus::Pending { .. }) => QueryResponseStatus::Pending,
+			_ => QueryResponseStatus::NotFound,
```

Comment:
In this case we are expecting a response for a query, not a version notification, that's why I mapped it to `NotFound`.

I could take the `_` out, but I think I'd keep mapping the `QueryStatus::VersionNotifier` to `NotFound`. What do you think?

Reply:
this would mean this impl of XcmQueryHandler is not general purpose, its not relevant for version notifications.
and this changes the behaviour of initial take_response, instead of None (no response), the user gets NotFound.

Reply:
This does not change the behaviour of the initial `take_response`, it makes it more general by going from an `Option` to an enum with more variants.
We could add another variant to the return enum for `VersionNotifier` to make it more general.

Reply:
@muharem Tell me what you think about the new code